### ISSUE TITLE
How to call another model while a graph is executed in a tf.Session

### ISSUE THREAD

**Issue by iamxiaoyubei** [tensorflow/lingvo#108]
I trained two models: ASR Wpm model and Language model. When I passed a intermediate value of ASR model to Language model in order to predict the intermediate result, I got an error. Because tensorflow is not able to pass tensor into feed_dict. But I can only get the value which passed to language model while the ASR model is running. So this value can only be saved in the form of tensor. So, could you tell me how can I call language model in ASR model? I really hope that you can help me. Thank you!!

**Comment by jonathanasdf** [comment]
You can't use two separate inference graphs if you want to combine two models. You will need to fuse them together into a single graph.

**Comment by iamxiaoyubei** [comment]
I am confused about how to fuse two separate graph together into a single graph. Could you give me some tips?

**Comment by iamxiaoyubei** [comment]
What I did was call `getNextLogitsByPrevLogits` in the `ComputeLogitsWithLM` function to get next_logits.
```
  def getNextLogitsByPrevLogits(self, prev_id):
    checkpoint = tf.train.latest_checkpoint('/tmp/lm1b/librilog/train')
    params = model_registry.GetParams('lm.one_billion_wds.LibrispeechLMWordPiece', 'Test')
    inference_graph = inference_graph_exporter.InferenceGraphExporter.Export(params)
    pred = predictor.Predictor(inference_graph, checkpoint=checkpoint, device_type='gpu')
    next_logits = pred.Run(['next_logits'], id=prev_id)
    return next_logits

def ComputeLogitsWithLM(self, state, logits, is_eval=False):
    ...
    next_logits = self.getNextLogitsByPrevLogits(ids)
    ...
```
These codes will run once when the ASR inference graph is built. Then they are part of the graph. But since the `ids(prev_id)` is tensor, there is no way to pass it into pred.Run(which will pass it to feed_dict to run LM inference graph).

Therefore I really don't know how to fuse two graphs because I think the above is also a way to fuse two graphs. Besides, it is the only way I can come up with now(I am sorry for my ability). So, I hope you can help me.

**Comment by jonathanasdf** [comment]
The graph generated by InferenceGraphExporter is self contained and not actually merged into the graph you're currently building. Instead of using the model registry and inference_graph_exporter you need to actually integrate the lm into your model by:

1) configuring the lm param in the task params
https://github.com/tensorflow/lingvo/blob/2d05484a7d5d73db23f8a4b47d6d729b5e01fa6a/lingvo/tasks/asr/params/librispeech.py#L314

eg. `p.decoder.fusion.lm = CreateLmParams()`

2) then in `ComputeLogitsWithLM` you can use `self.lm.Logits(theta.lm, inputs, paddings)`
https://github.com/tensorflow/lingvo/blob/2d05484a7d5d73db23f8a4b47d6d729b5e01fa6a/lingvo/tasks/lm/layers.py#L68

or use FProp if you need the other return values too.

**Comment by iamxiaoyubei** [comment]
Thank you so much! However, through this way, must I train language model together with ASR? Can I also use the parameters values trained using LibrispeechLMWordPiece before? If yes, How could I load these parameters without InferenceGraphExporter? (I think it maybe load LM's checkpoint at the same time with loading ASR's checkpoint, but I am not sure how to do exactly)

**Comment by jonathanasdf** [comment]
You will need to write some script to update the variable names in the lm checkpoint and merge it together with the asr checkpoint. Tensorflow checkpoint format isn't too complicated. See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/inspect_checkpoint.py.

I just google "tensorflow rename variable checkpoint" and found 
https://gist.github.com/batzner/7c24802dd9c5e15870b4b56e22135c96

**Comment by iamxiaoyubei** [comment]
Thank you very much!! I will try it right away ; )

**Comment by iamxiaoyubei** [comment]
Through your way I successfully implemented shallow fusion. Thank you very much! Finally, I would like to ask what is the λ(which decides how many proportions to join the LM) that you generally set?
![image](https://user-images.githubusercontent.com/21981293/60258287-72b81c00-9907-11e9-982b-b21ef7c7fbec.png)

**Comment by anjuli** [comment]
lambda is typically between 0.1 and 0.3. We do some tuning on a dev set to find the right value.

See third paragraph of section 4.1.2 as well as second paragraph of section 4.2.2 here:
https://arxiv.org/pdf/1807.10857.pdf

You may also find these references informative if you're using Librispeech:
https://arxiv.org/pdf/1805.03294.pdf
https://arxiv.org/pdf/1902.01955.pdf

**Comment by iamxiaoyubei** [comment]
Thank you!!!

**Comment by zh794390558** [comment]
watch this.
